src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -279,6 +286,37 @@ public function addClassResource(\ReflectionClass $class)
     }
 
     /**
+     * Checks whether the requested class, interface or trait exists and registers the result for resource tracking.
+     *
+     * @param string $class
+     *
+     * @return bool
+     *
+     * @final
+     */
+    public function classExists($class)
+    {
+        if (isset($this->classReflectors[$class])) {
+            return (bool) $this->classReflectors[$class];
+        }
+        $exists = class_exists($class) || interface_exists($class, false) || trait_exists($class, false);
+
+        if ($this->trackResources) {
+            if (!$exists) {
+                $this->addResource(new ClassExistenceResource($class));
+                $this->classReflectors[$class] = false;
+            } else {
+                $class = $this->classReflectors[$class] = new \ReflectionClass($class);
```

Comment:
that could be an issue, but the code in #21419 is free from it, see
https://github.com/symfony/symfony/pull/21419/files#diff-a10c9afd8feaccd1ff66ffd54e2835bcR356

The reflection resource is always added, even if the cache is hit. This creates duplicate resources, but they are free, and deduplicated later on so not an issue.